Implement test script for US 307 - Auto_Admin is able to edit user information

honganh1206 opened the pull request:
https://dev.azure.com/RookiesBatchVI/RookiesBatch6Group4/_workitems/edit/307

All test scripts for US 307 to US 307 ran successfully (Dec 11, 2022 17:51):
![image](https://user-images.githubusercontent.com/79431746/206899435-ca621ec4-f4c8-48d1-b81c-ce9fb017a22d.png)

Test script for US 307:
![image](https://user-images.githubusercontent.com/79431746/206899455-e2a92f4b-81b5-41ba-a683-08bd5a4d6de5.png)

honganh1206 committed:
Fix locator naming in in CreateUserPage + ManageUserPage

honganh1206 committed:
Update TearDown in WebTestSetup

Update code to check if user created with API can be disabled (with GET request)

honganh1206 committed:
Fix locators + Rename folder

Renamed ActualData folder to DataFromUI
Fixed locators in ManageUserPage
Wrote sample test cases to give example of how to sort for Minh

Minh-Tester reviewed (approved):
Look good to me!

daipn1995 commented on AssetManagementTestProject/PageObj/ManageUserPage.cs:
```diff
@@ -1,71 +1,58 @@
+    public void SortUser(string sortType)
+    {
+        Click(sortType);
+        Thread.Sleep(5000);
```
Never use Thread sleep if you can => Try to wait for the loading icon disappear.
the hard sleep time should have a function to cover. For examples, waitForSmtAppear

honganh1206 replied:
Sorry for deleting the comment that explains why I used thread sleep here. I have tried to use explicit wait for some locators such as first row of the table, edit button,... but somehow it always stop the sorting. 
So far I have resorted to using thread sleep, but when I find an alternative I will update ASAP. I will make a seperate method for thread sleep too

daipn1995 commented on AssetManagementTestProject/TestCases/US307_EditUserTest.cs:
```diff
@@ -0,0 +1,88 @@
+        LoginPage?.Login(Constant.ADMIN_USERNAME_HN, Constant.ADMIN_PASSWORD);
+        DriverBaseAction?.WaitToBeVisible(HomePage.HeaderHomePage);
+        DriverBaseAction?.Click(MenuBarLeft.BtnManageUserInMenu);
+        ManageUserPage = new ManageUserPage();
```
Init Pages with the same way on NUnitWebTestSetUp

honganh1206 replied:
Done

daipn1995 commented on AssetManagementTestProject/TestCases/US307_EditUserTest.cs:
```diff
@@ -0,0 +1,88 @@
+[TestFixture]
+public class US307_EditUserTest : NUnitWebTestSetup
+{    
+    protected ManageUserPage? ManageUserPage; 
+    protected EditUserInfoPage? EditUserInfoPage;
+    protected DetailedUserInfoPage? DetailedUserInfoPage;
+    protected ViewUserDAO.ViewDetailedUser? DetailedUserInfoFromUI;
+    protected GetUserDAO.GetUserResponse? NewUserAfterEdit;
```
Can you move them into the parent class?

honganh1206 replied:
Done

daipn1995 commented on CoreFramework/DriverCore/WebDriverAction.cs:
```diff
@@ -448,6 +448,32 @@ public void AssertEquals(object actual, object expected)
+        }
+        catch(Exception excep)
+        {
+            HtmlReport.Fail("List is not sorted in ascending order");
```
print the list of strings.
If fail, you can know exact current order

honganh1206 replied:
Done

daipn1995 commented on CoreFramework/DriverCore/WebDriverAction.cs:
```diff
@@ -448,6 +448,32 @@ public void AssertEquals(object actual, object expected)
             throw excep;
         }
     }
+    public void AssertListAscending(List<string> list)
+    {
+        try
+        {
+            list.Should().BeInAscendingOrder();
+            HtmlReport.Pass("List is sorted in ascending order");
+        }
```
Is this really needed to write this function?
if you try to use the syntax simply `list.Should().BeInAscendingOrder();`
If it's failed, then after the test is done, it will add the Pass/Fail log for HTML report automatically

honganh1206 replied:
This is a sample method I wrote for Minh so he can finish his test script for US 306, so I will notify him later. Thank you for your suggestion

honganh1206 replied:
Done

daipn1995 commented on AssetManagementTestProject/TestCases/US307_EditUserTest.cs:
```diff
@@ -0,0 +1,88 @@
+        DetailedUserInfoPage detailedUserInfo = new DetailedUserInfoPage();
+        DetailedUserInfoFromUI = detailedUserInfo.ReturnDetailedUserInfoFromUI();
+        // Update new user info with GET request
+        NewUserAfterEdit = APIService?.ReturnSelectedUser(NewUser.Data.Id.ToString(), Token);
```
This is incorrect.
If you are trying to edit a user, then you have to create the expected data by yourself.
IF you get data from API to be the expected data, THEN you're verifying the database with the database.
It's always true except the convert function from FrontEnd have some issues.

honganh1206 replied:
Done

honganh1206 committed:
Resolve most comments

Moved properties to WebTestSetup
Adjusted locators for Edit User
Created expected data for verification
Added Wait method for querying result

honganh1206 committed:
Update properties in WebTestSetup
